File: Daily/PY/Leetcode65-有效数字.py
class Solution:
    def ispurenumber(self,s,must_be_int):
        if not s:
            return False
        has_dot,has_digit = False,False

        st = 1 if s[0] == '+' or s[0] == '-' else 0

        for i in range(st,len(s)):
            if s[i] == '.':
                if must_be_int:
                    return False
                if has_dot:
                    return False
                has_dot = True
            elif s[i].isdigit():
                has_digit = True
            else:
                return False
        
        return has_digit
    
    def isNumber(self, s: str) -> bool:
        eps = s.find('e')
        Eps = s.find('E')

        if (eps != -1):
            return self.ispurenumber(s[:eps],False) and self.ispurenumber(s[eps+1:],True)

        if (Eps != -1):
            return self.ispurenumber(s[:Eps],False) and self.ispurenumber(s[Eps+ 1:],True)
    
        return self.ispurenumber(s,False)

File: Daily/Rust/Leetcode65-有效数字.rs
impl Solution {
    pub fn is_number(s: String) -> bool {
        let eps = s.find('e');
        let Eps = s.find('E');
        fn ispurenumber(s:&str,mustbeint:bool) -> bool {
            let (mut hasdigit,mut hasdot) = (false,false);
            let st = if s.chars().nth(0) == Some('+') || s.chars().nth(0) == Some('-'){1} else {0};
            for i in st..s.len(){
                if s.chars().nth(i).unwrap() == '.'{
                    if mustbeint{
                        return false;
                    }
                    if hasdot{
                        return false;
                    }
                    hasdot = true;
                }else if s.chars().nth(i).unwrap().is_digit(10){
                    hasdigit = true;
                }else{
                    return false;
                }
            }
            hasdigit
        }
        if let Some(i) = eps{
            return ispurenumber(&s[0..i],false) && ispurenumber(&s[i+1..],true);
        }
        if let Some(i) = Eps{
            return ispurenumber(&s[0..i],false) && ispurenumber(&s[i+1..],true);
        }
        ispurenumber(&s,false)
    } 
}